keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,1128 @@
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Standard Heuristic Implementation (kube-proxy dataplane)
+
+kube-proxy based dataplane (along with EndpointSlice controller, within
+kube-controller-manager as the control plane) will support the three standard
+routing preferences (`Default`, `PreferEqualSpread`, `PreferZone`).
+
+#### `Default` and `PreferEqualSpread`
+* Initially, kube-proxy will treat the `Default` preference the same as
+  `PreferEqualSpread`
+* This leverages existing implementation, requiring no major changes.
+
+#### `PreferZone`
+* This preference will be implemented by the use of Hints within EndpointSlices.
+* We already use Hints to implement `service.kubernetes.io/topology-mode=Auto`
+  Similarly, we’ll use the same Hints within the EndpointSlice to implement the
+  PreferZone heuristic – the hints will match the zone of the endpoint itself.
+* While it may seem redundant to populate the hints here since kube-proxy can
+  already derive the zone hint from the endpoints zone (as they would be the
+  same), we will still use this for implementation simply because of the reason
+  that it’s easier to implement and provides a better design. Consider an
+  alternative implementation where kube-proxy reads
+  `routingPreference=PreferZone` field and then constructs the route rules
+  accordingly. This means some extra logic needs to be baked into the kube-proxy
+  which could have just as easily been implemented by an already existing
+  extensibility mechanism (i.e. EndpointSlice hints)
+
+Although this is not an explicit design goal, an implication of the above
+implementation choice means that:
+* The control plane (kube-controller-manager in this case) is only concerned
+  with the field `routingPreference` and populates EndpointSlice hints based
+  on its value 
+* The data plane (kube-proxy in this case) is only concerned with the Hints
+  populated in the EndpointSlice and the fields `internal/externalTrafficPolicy`
+  to make routing decisions.
+* Neither the control plane nor the data plane looks at the others field.
+
+NOTE: The fact that EndpointSlice hints are not expected to be implemented by
+  all data planes is not of concern here, because `routingPreference` is anyways
+  supposed to be implementation dependent. Since we know that the kube-proxy
+  data plane will respect the Hints, it’s sufficient for the kube-proxy and
+  kube-controller-manager based implementations to say that we do support the
+  standard heuristics.
+
+### Status Reporting
+
+To provide clear status updates about the routing preferences to the user, the
+EndpointSlice controller (which is acting as the control plane) will update the
+Service status with the following conditions (inspired by Gateway API conditions)
+
+* RoutingPreferenceAccepted
```

Comment:
I think the only relevant component that could reliably set status on a Service here is the EndpointSlice controller. Given that, I think these are the states we'd want to be able to represent:

1. The EndpointSlice controller sees a value that is not domain-prefixed on a Service and does not know how to handle it
2. The EndpointSlice controller sees a value that it recognizes but can't populate hints due to some other factor (common with the current proportional approach)

It is likely worth representing the positive polarity versions of those as well:
1. The EndpointSlice controller sees a value that it recognizes and has populated hints or determined that it does not need to for the heuristic (Default and PreferEqualSpread would fall into the latter category as proposed)

I think the first item in both of those lists could be handled by this `RoutingPreferenceAccepted` condition that was true or false. Maybe it should be set to "Unknown" for unrecognized/domain-prefixed values?

I think the second item (error populating hints) should probably only appear in the error case. I think your suggestion of `RoutingPreferenceProgrammed` is reasonable here, but since this is very specific to hints, we could instead use something more specific like `HintsPopulated`.

Reply:
Thanks for giving this a detailed thought, Rob! I mostly have the same thoughts as you on the matter, with the exception of the topic regarding `RoutingPreferenceProgrammed` v/s `HintsPopulated` as the condition name. I was thinking that "HintsPopulated" would be a _reason_ under the `RoutingPreferenceProgrammed` condition. Here's what I had in mind:

* From the user's perspective, the top priority is whether their routing preference has been "somewhat" sent to the data plane. `RoutingPreferenceProgrammed` reflects this directly, while the use of hints may be thought of as an implementation detail (which we will convey through the condition's reason)

* Another reason might spur the debate of being generic vs _YAGNI_ :). But given the previous advantage, in this case I was leaning towards being generic which allows non-hint based implementations to reuse the same condition.

Reply:
Yep, that makes sense, I think that's a reasonable approach here.

Reply:
I think the primary reason we needed status on the previous iteration of topology was that there was so much confusion about when it was or was not applied. I'd recommend scrapping status from this stage of the KEP. We can choose to add it in the future if feedback suggests it would be helpful, but I'm hopeful that this will end up being closer to any other Service field where it's not really necessary.